### PR TITLE
Ensure dialog class is applied only once

### LINKED ISSUE #8241: Dialog: Class is applied twice
### Environment Information 
 
- __Package version(s)__: 6.145.0+

### Please provide a reproduction of the bug in a codepen:
 
https://codepen.io/erichdev/pen/rRmagV
 
#### Actual behavior:
 
Not sure if this is intended behavior, but when a className is passed to `Dialog`, it is applied twice:

`<Dialog className='XXXXXXXXX'  />` results in:

```html
<div class="ms-Layer ms-Layer--fixed ms-Dialog XXXXXXXXX root-62" data-portal-element="true">
  <div class="ms-Fabric ms-Layer-content content-65">
    <div role="alertdialog" aria-modal="true" aria-labelledby="Dialog4-title" aria-describedby="Dialog4-subText" style="outline: none;">
      <div class="ms-Modal is-open ms-Dialog XXXXXXXXX root-64">
```
With 6.144.0 and below, this is the result:
```html
<div class="ms-Layer ms-Layer--fixed root-61" data-portal-element="true">
  <div class="ms-Fabric ms-Layer-content content-64">
    <div role="alertdialog" aria-modal="true" aria-labelledby="Dialog4-title" aria-describedby="Dialog4-subText" style="outline: none;">
      <div class="ms-Modal is-open ms-Dialog XXXXXXXXX root-63">
```
I blieve it is the result of this PR: https://github.com/OfficeDev/office-ui-fabric-react/pull/8065
But I don't know if this is intended behavior for the sticky feature to work.

#### Priorities and help requested:
 
Are you willing to submit a PR to fix? No
 
Requested priority: Low
 
Products/sites affected: SPO. This affected a test, but no other effects as far as I can tell.

### PR DESCRIPTION
**Pull request by aftab-hassan**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #8241 
- [x] Include a change request file using `$ npm run change`

#### Description of changes
Behavior reported in issue is a side effect of this [PR](https://github.com/OfficeDev/office-ui-fabric-react/pull/8065)
The modal `className` was getting incorrectly applied to the layer. Now using `layerProps.ClassName` instead of `this.props.className`. Thanks @jspurlin for suggesting the fix. 

#### Focus areas to test
Adding a `className` to a `Dialog` and ensuring that the `className` is applied only once

Before fix
class-name-before-fix

After fix
class-name-after-fix

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/8273)

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/Dialog/examples/Dialog.Basic.Example.tsx [hunk 1]
```diff
@@ -31,6 +31,7 @@ export class DialogBasicExample extends React.Component<{}, IDialogBasicExampleS
         <Dialog
           hidden={this.state.hideDialog}
           onDismiss={this._closeDialog}
+          className="XXXXXXXXX"
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/Dialog/examples/Dialog.Basic.Example.tsx]: If you want to do this to generate new snapshot tests for verification I suggest using a less jarring `className` value such as `"foo"`

Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/Dialog/examples/Dialog.Basic.Example.tsx]: Though I think since we're not applying any styles, we should simply move this to be encapsulated in a unit test which verifies a `className` on `Dialog` is only applied once.

If we add it to an example and generate snapshots that is a valid way of testing. However, developers frequently copy our examples to bootstrap their development. Having an artifact like an empty `className` would simply be more cognitive overload for them to deal with.

Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/Dialog/examples/Dialog.Basic.Example.tsx]: It should be a straight-forward unit test to add...

File: packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts [hunk 2]
```diff
@@ -118,6 +118,8 @@ export type IModalStyleProps = Required<Pick<IModalProps, 'theme'>> &
     hasBeenOpened?: boolean;
     /** Positioning of modal on first render */
     modalRectangleTop?: number;
+    /** Props for layer element */
+    layerProps?: ILayerProps;
```
Comment (KevinTCoughlin) [inline packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: I find it surprising that CI is passing without having run `npm run update-api`. This seems to me like a type change to `IModalStyleProps` which is exported publicly 🤔

Comment (JasonGore) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: [This](https://github.com/OfficeDev/office-ui-fabric-react/blob/7cf240e434ebf5b9961be95ec6bb2d2503b4907b/packages/office-ui-fabric-react/etc/office-ui-fabric-react.api.ts#L13177) is concerning. It's not expanded in the API file and rather is `WARNING: Unsupported export` 😕

File: common/changes/office-ui-fabric-react/fix-dialog-class-twice_2019-03-12-05-13.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "office-ui-fabric-react",
+      "comment": "Modal-Ensure dialog class is applied only once",
```
Comment (dzearing) [inline common/changes/office-ui-fabric-react/fix-dialog-class-twice_2019-03-12-05-13.json]: nit, use colon:

`Modal: Ensure that dialog class name is only applied once.`

File: packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts (hunk 2 above)
Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: @adimittal91 @natalieethell This has been a missing feature in api-extractor that Pete was going to fix for a long time. Do you know what happened with that?

Comment (natalieethell) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: All of the "WARNING: Unsupported export" warnings should go away when we upgrade to api-extractor 7 (which is still in beta, but should be officially released soon). 

`IModalStyleProps`, specifically, looks like this in the new api.ts file generated from api-extractor 7:

```tsx
// @public (undocumented)
declare type IModalStyleProps = Required<Pick<IModalProps, 'theme'>> & Pick<IModalProps, 'className' | 'containerClassName' | 'scrollableContentClassName' | 'topOffsetFixed' | 'isModeless'> & {
    isOpen?: boolean;
    isVisible?: boolean;
    hasBeenOpened?: boolean;
    modalRectangleTop?: number;
};
```

**Comment by aftab-hassan** [comment]
@JasonGore could you help review?

File: packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts (hunk 2 above)
Comment (JasonGore) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: I should have mentioned this earlier but can you call this prop `layer` instead of `layerProps`? It will line up with the `layer` style section naming and with Slots functionality we're introduction with Fabric 7, and will help ensure this component is backwards compatible. Thanks!

Comment (KevinTCoughlin) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: > I should have mentioned this earlier but can you call this prop `layer` instead of `layerProps`? It will line up with the `layer` style section naming and with Slots functionality we're introducting with Fabric 7, and will help ensure this component is backwards compatible. Thanks!

@aftab-hassan can you make the above change re: `layer` vs `layerProps` prior to us merging this?

File: packages/office-ui-fabric-react/src/components/Modal/Modal.base.tsx [hunk 4]
```diff
@@ -140,21 +140,14 @@ export class ModalBase extends BaseComponent<IModalProps, IDialogState> implemen
...
-      ? this.props.className
-        ? `${this.props.className} ${classNames.layer}`
-        : classNames.layer
-      : this.props.className;
-
     const mergedLayerProps = {
       ...DefaultLayerProps,
       ...this.props.layerProps,
       onLayerDidMount: layerProps && layerProps.onLayerDidMount ? layerProps.onLayerDidMount : onLayerDidMount,
-      className: layerClassName,
```
Comment (jspurlin) [inline packages/office-ui-fabric-react/src/components/Modal/Modal.base.tsx]: > className: layerClassName, [](start = 5, length = 27)

doesn't classNames.layer still need to be be passed down to the layer?

You could make the getClassNames always set layerProps && layerProps.className on layer and then add the additional styling if it isModeless. If you did that you could just always say className: classNames.layer here.

Comment (aftab-hassan) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.base.tsx]: @jspurlin thanks for the comment. That is the implementation I had in this [commit](https://github.com/OfficeDev/office-ui-fabric-react/pull/8273/commits/7e0229812231db1d891af20f8f8b1c6df33bf87a). But got a suggestion about doing the styling in the styles file, `Modal.styles.ts`. Please advise on whether the changes there look okay.

Comment (jspurlin) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.base.tsx]: So you are setting the styling in Modal.styles.ts (which updates the className so that it includes the original classname plus potentially the new classname) but you still need to set className to classNames.layer so that it can use that updated value

File: packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts (hunk 2 above)
Comment (aftab-hassan) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: @JasonGore @KevinTCoughlin I have renamed `layerProps` to `layer`. I feel like I have misunderstood the ask. There is now a `layer` prop of type `IStyle` as well as `ILayerProps`, in the `Modal.types.ts` file, which makes it difficult to read IMO.

File: packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts [hunk 5]
```diff
@@ -55,7 +55,7 @@ export interface IModalProps extends React.ClassAttributes<ModalBase>, IWithResp
   /**
    * Props to be passed through to Layer
    */
-  layerProps?: ILayerProps;
+  layer?: ILayerProps;
```
Comment (jspurlin) [inline packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: Why are you changing the name of this prop, seems completely unrelated to what this PR is trying to fix (and it's a breaking change)

Comment (natalieethell) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: @JasonGore asked the following in https://github.com/OfficeDev/office-ui-fabric-react/pull/8273#discussion_r266679679:

> I should have mentioned this earlier but can you call this prop `layer` instead of `layerProps`? It will line up with the `layer` style section naming and with Slots functionality we're introducting with Fabric 7, and will help ensure this component is backwards compatible. Thanks! 

@aftab-hassan if you rename the prop, you should keep `layerProps` and mark it as deprecated (use @deprecated), and then make sure that both the new `layer` and the old `layerProps` work in the same way. This is to maintain backwards compatibility.

Comment (natalieethell) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: Here's an example of how we do this for `collapsed` and `isCollapsed` in SearchBox:

https://github.com/OfficeDev/office-ui-fabric-react/blob/78a6157e3caa57db9634d0b11d245b3187f573e6/packages/office-ui-fabric-react/src/components/SearchBox/SearchBox.types.ts#L26-L35

https://github.com/OfficeDev/office-ui-fabric-react/blob/78a6157e3caa57db9634d0b11d245b3187f573e6/packages/office-ui-fabric-react/src/components/SearchBox/SearchBox.base.tsx#L29-L32

https://github.com/OfficeDev/office-ui-fabric-react/blob/78a6157e3caa57db9634d0b11d245b3187f573e6/packages/office-ui-fabric-react/src/components/SearchBox/SearchBox.base.tsx#L69

Comment (JasonGore) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: I'm sorry, @jspurlin is completely right. For some reason I thought `layerProps` was a new prop added in this PR. Don't worry about renaming or deprecation as part of this PR. I'm not sure how I misintepreted that. 😕

File: packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts [hunk 6]
```diff
@@ -118,6 +118,8 @@ export type IModalStyleProps = Required<Pick<IModalProps, 'theme'>> &
     hasBeenOpened?: boolean;
     /** Positioning of modal on first render */
     modalRectangleTop?: number;
+    /** Props for layer element */
+    layer?: ILayerProps;
```
Comment (JasonGore) [inline packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: This should be added to the `Pick` list above instead of being duplicated.

File: packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts (hunk 5 above)
Comment (aftab-hassan) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: Thanks for the sample @natalieethell 

I'll revert the naming @JasonGore

File: packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts (hunk 2 above)
Comment (aftab-hassan) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.types.ts]: Marking this comment as `Resolved`. I reverted the naming based on this comment https://github.com/OfficeDev/office-ui-fabric-react/pull/8273#discussion_r268243178

File: packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts [hunk 7]
```diff
@@ -22,7 +22,8 @@ export const getStyles = (props: IModalStyleProps): IModalStyles => {
     modalRectangleTop,
     theme,
     topOffsetFixed,
-    isModeless
+    isModeless,
+    layerProps
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts]: Please pass in the exact thing you need, rather than objects. It will cache better.

File: packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts [hunk 8]
```diff
@@ -83,6 +84,7 @@ export const getStyles = (props: IModalStyleProps): IModalStyles => {
       scrollableContentClassName
     ],
     layer: isModeless && [
+      layerProps && layerProps.className,
```
Comment (dzearing) [inline packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts]: actually I'm confused why you have to do this. The classname SHOULD be applied by the layer if the layerprops are being passed along.

**Comment by dzearing** [comment]
@aftab-hassan This has fallen off the radar. Please close the PR and resolve, or resolve today. Thanks

File: packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts (hunk 7 above)
Comment (aftab-hassan) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts]: @dzearing I am using `layerProps && layerProps.className,` below, so all of `layerProps` is used. 

> The classname SHOULD be applied by the layer if the layerprops are being passed along.

If I have understood your [comment](https://github.com/OfficeDev/office-ui-fabric-react/pull/8273/files/50853adc8e908f88d18a71dfc8dec05f2afac772#r271009994) correctly, `layerProps` I believe is still needed. Please correct me if I am wrong.

File: packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts (hunk 8 above)
Comment (aftab-hassan) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts]: @dzearing Is your recommendation to remove `layerProps && layerProps.className,` and so avoid passing in the `layerProps` as you mention [above](https://github.com/OfficeDev/office-ui-fabric-react/pull/8273#discussion_r271009707)?

Comment (aftab-hassan) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts]: > I believe this is needed so that `layerProps.className` will continue to get passed down to the layer while also adding the needed styling (below). Without this the `layerProps.className` will be completely overridden by the below styling when modeless
> 
> In reply to: [271020723](https://github.com/OfficeDev/office-ui-fabric-react/pull/8273#discussion_r271020723) [](ancestors = 271020723)

@dzearing, if we do keep the `layerProps && layerProps.className`, then, we need the `layerProps` prop. Please advise on this, and if you have any other feedback. 

@JasonGore if you had any other comments too.

File: packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts (hunk 7 above)
Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts]: you just need layerClassName

File: packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts (hunk 8 above)
Comment (aftab-hassan) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts]: @dzearing does this resolve your comment? Please let me know how we can complete this PR.

File: packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts (hunk 7 above)
Comment (dzearing) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts]: You do not NEED all of the layerProps. You just need the layerClassName. This function (`getClassNames`) should be taking immutable objects as input, so that we can cache results more efficiently. You are passing in a constantly mutating object here which would break caching. Please pull out exactly what you need.

Comment (aftab-hassan) [reply packages/office-ui-fabric-react/src/components/Modal/Modal.styles.ts]: okay, shall submit a fix.

**Comment by aftab-hassan** [comment]
@dzearing @jspurlin to be fixed today.

**Comment by dzearing** [comment]
The screener issue seems confusing and non related. Requeuing.

**Comment by apneer** [comment]
Component Perf Results:
  <table>
  <tr>
    <th>Scenario</th>
    <th>Current Avg Total (ms)</th>
    <th>New Avg Total (ms)</th>
    <th>Current Avg Per Item (ms)</th>
    <th>New Avg Per Item (ms)</th>
  </tr><tr>
            <td>PrimaryButton</td>
            <td>105.968</td>
            <td>107.620</td>
            <td>1.060</td>
            <td>1.076</td>
           </tr>  
<tr>
            <td>BaseButton</td>
            <td>36.020</td>
            <td>37.594</td>
            <td>0.360</td>
            <td>0.376</td>
           </tr>  
<tr>
            <td>NewButton</td>
            <td>110.035</td>
            <td>112.608</td>
            <td>1.100</td>
            <td>1.126</td>
           </tr>  
<tr>
            <td>button</td>
            <td>5.414</td>
            <td>5.082</td>
            <td>0.054</td>
            <td>0.051</td>
           </tr>  
<tr>
            <td>DetailsRows without styles</td>
            <td>202.062</td>
            <td>203.799</td>
            <td>2.021</td>
            <td>2.038</td>
           </tr>  
<tr>
            <td>DetailsRows</td>
            <td>209.090</td>
            <td>212.003</td>
            <td>2.091</td>
            <td>2.120</td>
           </tr>  
<tr>
            <td>Toggles</td>
            <td>51.657</td>
            <td>50.824</td>
            <td>0.516</td>
            <td>0.508</td>
           </tr>  
<tr>
            <td>NewToggle</td>
            <td>72.841</td>
            <td>68.693</td>
            <td>0.729</td>
            <td>0.687</td>
           </tr>  </table><div id="perfComment9423"/>

**Comment by aftab-hassan** [comment]
@JasonGore could you please take a look at this one?
